### PR TITLE
Cilium: Encryption, remove ingress IPsec rules when no longer needed

### LINKED ISSUE #15216: Cilium: Encryption, remove ingress IPsec rules when no longer needed
When encryption is disabled we remove xfrm entries for remote nodes,
```
 NodeConfigurationChanged()
   n.removeEncryptRules()
   ipsec.DeleteXfrm()
```
This stops the node from sending encrypted traffic, because it no longer has rules/xfrm-state/policy to do this. However, we have to leave the ingress decrypt rules in place because its unclear when the remote nodes will be updated and we don't want to drop traffic if a remote node sends us an encrypted traffic. For example in large deployments a rolling update can take some time and so its not possible to just break traffic. Worse it can cause more than a disruption if we lose connection with critical services, apiService, kvstore, etc.

The current solution is to simply leave the decrypt rules in place. They will only ever be used if an encrypted packet is received then after all nodes have transitioned should never be used. This is mostly harmless, except for in a corner case here,
 
 1. enable encrypt with keyID=1 and key=foo
 2. disable encrypt with keyID=1
 3. (re)enable encrypt with keyID=1 and key=bar
 
When step3 is run and we (re)enable encrypt we attempt to install a new decrypt rule with the same ID but different key. This will collide with the left-over key from step1 and install will fail leaving a warning in the logs. To work-around above case we recommend setting keyID to some other value, replacing step3 with the following,

 3'. (re)enable encrypt with ikeyID=2 and key=bar
 
 The new keyID will get installed correctly and the old keyID will be removed after a timeout.
 
 This feature request is to delete keyID=1 in step2 after we recognize all nodes have been updated. We can determine this by reviewing the ipCache. If the key is no longer listed in the ipcache then all nodes have sent us an update with the new key state, in this case no key because we have deleted the key.
 
 This will remove the above corner case and also has the nice side benefit of removing stale rules in the policy/state of ipsec.

### PR DESCRIPTION
**Pull request by weirdwiz**
This commit checks if a key exists in ipCache, and if it doesn't then removes it's corresponding Xfrm entries.

Fixes: #15216

### CODE REVIEW

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 1]
```diff
@@ -604,3 +604,8 @@ func DeleteIPsecEncryptRoute() {
 		}
 	}
 }
+
+func DeleteIPSecXfrmSpi(spi uint8) {
+	time.Sleep(linux_defaults.IPsecKeyDeleteDelay)
+	ipsecDeleteXfrmSpi(spi)
+}
```
Comment (weirdwiz) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:611]: To delete the rules, I'm using ipsecDeleteXfrmSpi() but that function isn't exported, so I've created a wrapper function to call it.

File: pkg/datapath/linux/node.go [hunk 2]
```diff
@@ -897,6 +899,25 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 		oldKey = oldNode.EncryptionKey
 	}
 
+	if !n.nodeConfig.EnableIPSec {
+		bpfIPCacheList := make(map[string][]string)
+		if err := ipcache.IPCache.Dump(bpfIPCacheList); err != nil {
+			// TODO: Write error
+		}
```
Comment (weirdwiz) [inline pkg/datapath/linux/node.go]: To check values from IPCache I've used IPCache.Dump() from github.com/cilium/cilium/pkg/maps/ipcache . Would this be the correct functions to use?

File: pkg/datapath/linux/node.go [hunk 3]
```diff
@@ -897,6 +899,25 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 		oldKey = oldNode.EncryptionKey
 	}
 
+	if !n.nodeConfig.EnableIPSec {
+		bpfIPCacheList := make(map[string][]string)
+		if err := ipcache.IPCache.Dump(bpfIPCacheList); err != nil {
```
Comment (weirdwiz) [inline pkg/datapath/linux/node.go:932]: I'm checking for the key in ipcache in the function `nodeUpdate`, is this the correct function to check? or should I shift this to `NodeConfigurationChanged()`

**Comment by jibi** [comment]
Thanks for the PR @weirdwiz!
Could you rebase your branch as there's a conflict in `node.go` please?

**Comment by weirdwiz** [comment]
Rebased

**Pull request closed without merging**